portable/ThirdParty/GCC/RP2040/port.c

```diff
@@ -0,0 +1,879 @@
+        }
+        else
+        {
+            configASSERT( portIS_FREE_RTOS_CORE() );
+            configASSERT( pxYieldSpinLock == NULL );
+
+            TickType_t uxTicksToWait = prvGetTicksToWaitBefore( uxUntil );
+            if( uxTicksToWait )
+            {
+                /* We want to hold the lock until the event bits have been set; since interrupts are currently disabled
+                 * by the spinlock, we can defer until portENABLE_INTERRUPTS is called which is always called when
+                 * the scheduler is unlocked during this call */
+                configASSERT(pxLock->spin_lock);
+                pxYieldSpinLock = pxLock->spin_lock;
+                ulYieldSpinLockSaveValue = ulSave;
+                xEventGroupWaitBits( xEventGroup,
+                                     prvGetEventGroupBit(pxLock->spin_lock), pdTRUE,
+                                     pdFALSE, uxTicksToWait );
+                /* sanity check that interrupts were disabled, then re-enabled during the call, which will have
+                 * taken care of the yield */
+                configASSERT( pxYieldSpinLock == NULL );
+            }
+            else
+            {
+                spin_unlock( pxLock->spin_lock, ulSave );
+            }
+            if ( time_reached( uxUntil ) )
+            {
+                return true;
+            }
+            else
+            {
+                /* We do not want to hog the core */
+                portYIELD();
+                /* We aren't sure if we've reached the timeout yet; the caller will check */
+                return false;
+            }
+        }
+    }
+
+    #if ( configSUPPORT_PICO_SYNC_INTEROP == 1)
+        /* runs before main */
+        static void __attribute__((constructor)) prvRuntimeInitializer( void )
+        {
+            /* This must be done even before the scheduler is started, as the spin lock
+             * is used by the overrides of the SDK wait/notify primitives */
+            #if ( LIB_PICO_MULTICORE == 1 )
+                pxCrossCoreSpinLock = spin_lock_instance( next_striped_spin_lock_num() );
+            #endif /* portRUNNING_ON_BOTH_CORES */
+
+            /* The event group is not used prior to scheduler init, but is initialized
+             * here to since it logically belongs with the spin lock */
+            #if ( configSUPPORT_STATIC_ALLOCATION == 1 )
+                xEventGroup = xEventGroupCreateStatic(&xStaticEventGroup);
+            #else
+                xEventGroup = xEventGroupCreate();
+            #endif /* configSUPPORT_STATIC_ALLOCATION */
+        }
+    #endif
+#endif /* configSUPPORT_PICO_SYNC_INTEROP */
+
+#if ( configSUPPORT_PICO_TIME_INTEROP == 1 )
+    void xPortSyncInternalYieldUntilBefore( absolute_time_t t )
+    {
+        TickType_t uxTicksToWait = prvGetTicksToWaitBefore(t);
+        if( uxTicksToWait )
+        {
+            vTaskDelay(uxTicksToWait);
+        }
+    }
+#endif /* configSUPPORT_PICO_TIME_INTEROP */
```

Comment:
Please add a new line at the end of this file.

Reply:
ok